jax/_src/scipy/signal.py

```diff
@@ -146,3 +148,351 @@ def detrend(data, axis=-1, type='linear', bp=0, overwrite_data=None):
+    if input_length < win.shape[-1]:
+      raise ValueError('window is longer than input signal')
+    if nperseg is None:
+      nperseg = win.shape[0]
+    elif nperseg is not None:
+      if nperseg != win.shape[0]:
+        raise ValueError("value specified for nperseg is different"
+                          " from length of window")
+  return win, nperseg
+
+def _fft_helper(x, win, detrend_func, nperseg, noverlap, nfft, sides):
+  """Internal helper function for calculating windowed FFT.
+
+  This funciton is basically a jax re-implementation of
+  https://github.com/scipy/scipy/blob/47bb6febaa10658c72962b9615d5d5aa2513fa3a/scipy/signal/spectral.py#L1874
+  """
+  # Created strided array of data segments
+  if nperseg == 1 and noverlap == 0:
+    result = x[..., np.newaxis]
+  else:
+    step = nperseg - noverlap
+    *batch_shape, signal_length = x.shape
+    batch_shape = tuple(batch_shape)
+    x = x.reshape((int(np.prod(batch_shape)), signal_length))[..., np.newaxis]
+    result = jax.lax.conv_general_dilated_patches(
+        x, (nperseg,), (step,),
+        'VALID',
+        dimension_numbers=('NTC', 'OIT', 'NTC'))
+    result = result.reshape(batch_shape + result.shape[-2:])
+
+  # Detrend each data segment individually
+  result = detrend_func(result)
+
+  # Apply window by multiplication
+  result = win.reshape((1,) * len(batch_shape) + (1, nperseg)) * result
+
+  # Perform the fft. Acts on last axis by default. Zero-pads automatically
+  if sides == 'twosided':
+    func = jax.numpy.fft.fft
+  else:
+    result = result.real
+    func = jax.numpy.fft.rfft
+  result = func(result, n=nfft)
+
+  return result
+
+@_wraps(osp_signal.signaltools.odd_ext)
+def odd_ext(x, n, axis=-1):
+  """Redefined with jax.numpy for supporting auto-diff."""
+  def axis_slice(a, start=None, stop=None, step=None, axis=-1):
+    a_slice = [slice(None)] * a.ndim
+    a_slice[axis] = slice(start, stop, step)
+    b = a[tuple(a_slice)]
+    return b
+  if n < 1:
+    return x
+  if n > x.shape[axis] - 1:
+    raise ValueError(("The extension length n (%d) is too big. " +
+                      "It must not exceed x.shape[axis]-1, which is %d.")
+                      % (n, x.shape[axis] - 1))
+  left_end = axis_slice(x, start=0, stop=1, axis=axis)
+  left_ext = axis_slice(x, start=n, stop=0, step=-1, axis=axis)
+  right_end = axis_slice(x, start=-1, axis=axis)
+  right_ext = axis_slice(x, start=-2, stop=-(n + 2), step=-1, axis=axis)
+  ext = jnp.concatenate((2 * left_end - left_ext,
+                         x,
+                         2 * right_end - right_ext),
+                         axis=axis)
+  return ext
+
+
+_spectral_helper_mode = ('psd', 'stft')
+
+def _spectral_helper(x, y, fs=1.0, window='hann', nperseg=None, noverlap=None,
+                     nfft=None, detrend_type='constant', return_onesided=True,
+                     scaling='density', axis=-1, mode='psd', boundary=None,
+                     padded=False):
```

Comment:
Any numerical arguments that are required to be concrete should be checked here with `core._concrete_or_error`. I think this would include `nperseg`, `noverlap`, & `nfft` (maybe more?)

Reply:
Applied with adding a check for `axis`.